src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                    // flow control back pressure being applied (hasMoreData)
+
+                    stream.ResponseTrailers.SetReadOnly();
+                    stream.DecrementActiveClientStreamCount();
+
+                    // It is faster to write data and trailers together. Locking once reduces lock contention.
+                    flushResult = await WriteDataAndTrailersAsync(stream, buffer, producer.WriteHeaders, stream.ResponseTrailers);
+                }
+                else if (readResult.IsCompleted && producer.StreamEnded)
+                {
+                    if (buffer.Length != 0)
+                    {
+                        _log.LogCritical("{StreamId} observed an unexpected state where the streams output ended with data still remaining in the pipe.", stream.StreamId);
+                    }
+                    else
+                    {
+                        stream.DecrementActiveClientStreamCount();
+
+                        // Headers have already been written and there is no other content to write
+                        flushResult = await FlushAsync(stream, producer.WriteHeaders, outputAborter: null, cancellationToken: default);
+                    }
+                }
+                else
+                {
+                    var endStream = readResult.IsCompleted && !hasMoreData;
+
+                    if (endStream)
+                    {
+                        stream.DecrementActiveClientStreamCount();
+                    }
+
+                    flushResult = await WriteDataAsync(stream, buffer, buffer.Length, endStream, producer.WriteHeaders);
+                }
+
+                if (producer.IsTimingWrite)
+                {
+                    _timeoutControl.StopTimingWrite();
+                }
+
+                producer.WriteHeaders = false;
+
+                reader.AdvanceTo(buffer.End);
+
+                if ((readResult.IsCompleted && !hasMoreData) || readResult.IsCanceled)
+                {
+                    await reader.CompleteAsync();
+
+                    producer.CompleteResponse();
+                }
+                // We're not going to schedule this again if there's no remaining window.
+                // When the window update is sent, the producer will be re-queued if needed.
+                else if (hasMoreData)
+                {
+                    // We have no more connection window, put this producer in a queue waiting for it to
+                    // a window update to resume the connection.
+                    if (remainingConnection == 0)
+                    {
+                        // Mark the output as waiting for a window upate to resume writing (there's still data)
+                        producer.MarkWaitingForWindowUpdates(true);
+
+                        lock (_windowUpdateLock)
+                        {
+                            _waitingForMoreWindow.Enqueue(producer);
```

Comment:
If a stream used all of the available stream window and connection window at once, the stream could get enqueued at least twice into `_waitingForMoreWindow` without ever being dequeued if it received a stream window update before a connection window update. This can't be repeated because of the `wasDepleted` check in `UpdateWindow()`, but I think it would be safest to check that `producer._waitingForWindowUpdates` isn't already set.

Reply:
Handled by schedule